src/link/schema/index.ts

```
@@ -48,14 +49,21 @@ export class SchemaLink extends ApolloLink {
             ? this.context(operation)
             : this.context
         )
-      ).then(context => execute(
-        this.schema,
-        operation.query,
-        this.rootValue,
-        context,
-        operation.variables,
-        operation.operationName,
-      )).then(data => {
+      ).then(context => {
+        const validationErrors = validate(this.schema, operation.query);
+        if (validationErrors.length > 0) {
+          return { errors: validationErrors };
+        }
+
+        return execute(
+          this.schema,
+          operation.query,
+          this.rootValue,
+          context,
+          operation.variables,
+          operation.operationName,
+        )
```

Usually when you use `graphql` to execute a query, the query is validated against the schema. When you use the lower level `execute`, this validation is skipped and an empty object would be returned. The real error is swallowed.

We could theoretically use `graphql` which includes the validation, but then we have to `print` the schema again, since the function doesn't accept a parsed `DocumentNode`. Internally it then parses the query again.

Therefore I think the better solution is add the necessary validation here instead and continue to use `execute`.

See https://github.com/graphql/graphql-js/blob/3aad20bf1cccda59ccb8d855584097fcf7348fef/src/graphql.js#L189-L193

---

I agree with this reasoning!

Including the `graphql/validation/validate` dependency tree will increase bundle sizes, but only for apps using `@apollo/client/link/schema`—and of course one good solution is to move schema/execution/validation-related logic to a GraphQL server, instead of using this link.

---

Thanks for the review!

> one good solution is to move schema/execution/validation-related logic to a GraphQL server, instead of using this link.

Yep, I agree! At least for me the use cases for the schema link so far were to use it for SSR when the UI server also provides the GraphQL endpoint or for providing mock data in tests. In both cases bundle size shouldn't be an issue IMO 🙂.

If you're using the schema link client side, you're including the whole schema with all resolvers, which I guess already has quite a bundle size overhead …

---

@benjamn Also thanks for adding the changelog entry!

I just saw this wording:

> The schema link package (@apollo/client/link/schema) will now validate incoming queries against its client-side schema, and return errors as a GraphQL server would.

I'm wondering if the "client-side schema" and "server" part could be a bit misleading? Using the schema link implies that you're providing a local schema, regardless of if it's server or client side, right?

Maybe this could be an alternative?

> The schema link package (`@apollo/client/link/schema`) will now validate incoming queries against the schema before execution and therefore now has the same error handling capabilities as what a developer might be familiar with from a GraphQL endpoint or using `graphql` directly.

Not sure if it's too wordy? I'm also not a native English speaker, so there's likely something to be improved 🙂.

---

@amannn Improved the wording a bit in https://github.com/apollographql/apollo-client/pull/7015/commits/e27be1e31883c4a711206b1e7d1f179be3bbcc17.